internal/cryptokit/hash.go

```diff
@@ -58,3 +63,273 @@ func SHA512(p []byte) (sum [64]byte) {
 	C.SHA512(base(p), C.size_t(len(p)), base(sum[:]))
 	return
 }
+
+var (
+	MD5BlockSize    = int(C.MD5BlockSize())
+	MD5Size         = int(C.MD5Size())
+	SHA1BlockSize   = int(C.SHA1BlockSize())
+	SHA1Size        = int(C.SHA1Size())
+	SHA256BlockSize = int(C.SHA256BlockSize())
+	SHA256Size      = int(C.SHA256Size())
+	SHA384BlockSize = int(C.SHA384BlockSize())
+	SHA384Size      = int(C.SHA384Size())
+	SHA512BlockSize = int(C.SHA512BlockSize())
+	SHA512Size      = int(C.SHA512Size())
+)
+
+// cloneHash is an interface that defines a Clone method.
+//
+// hash.CloneHash will probably be added in Go 1.25, see https://golang.org/issue/69521,
+// but we need it now.
+type cloneHash interface {
+	hash.Hash
+	// Clone returns a separate Hash instance with the same state as h.
+	Clone() hash.Hash
+}
+
+var _ hash.Hash = (*evpHash)(nil)
+var _ cloneHash = (*evpHash)(nil)
+
+type evpHash struct {
+	pinner    runtime.Pinner
+	ptr       unsafe.Pointer
+	blockSize int
+	size      int
+
+	writeFunc func(p0 unsafe.Pointer, p1 *C.uint8_t, p2 C.int) C.int
+	sumFunc   func(p0 unsafe.Pointer, p1 *C.uint8_t)
+	resetFunc func(p0 unsafe.Pointer)
+	cloneFunc func(p0 unsafe.Pointer) (r1 unsafe.Pointer)
+	freeFunc  func(p0 unsafe.Pointer)
+}
+
+func newEVPHash(ptr unsafe.Pointer, blockSize, size int,
+	writeFunc func(p0 unsafe.Pointer, p1 *C.uint8_t, p2 C.int) C.int,
+	sumFunc func(p0 unsafe.Pointer, p1 *C.uint8_t),
+	resetFunc func(p0 unsafe.Pointer),
+	cloneFunc func(p0 unsafe.Pointer) (r1 unsafe.Pointer),
+	freeFunc func(p0 unsafe.Pointer)) *evpHash {
+	h := &evpHash{
+		ptr:       ptr,
+		blockSize: blockSize,
+		size:      size,
+		writeFunc: writeFunc,
+		sumFunc:   sumFunc,
+		resetFunc: resetFunc,
+		cloneFunc: cloneFunc,
+		freeFunc:  freeFunc,
+	}
+
+	runtime.SetFinalizer(h, (*evpHash).finalize)
+
+	return h
+}
+
+func (h *evpHash) finalize() {
+	if h.ptr != nil {
+		h.freeFunc(h.ptr)
+		h.ptr = nil
+	}
+}
+
+func (h *evpHash) Clone() hash.Hash {
+	if h.ptr == nil {
+		return nil
+	}
+	h.pinner.Pin(&h.ptr)
+	defer h.pinner.Unpin()
+
+	newHash := newEVPHash(h.cloneFunc(h.ptr), h.blockSize, h.size, h.writeFunc, h.sumFunc, h.resetFunc, h.cloneFunc, h.freeFunc)
+
+	return newHash
+}
+
+func (h *evpHash) Write(p []byte) (n int, err error) {
+	if len(p) > 0 {
+		h.pinner.Pin(&p[0])
+		defer h.pinner.Unpin()
+	}
+	if h.writeFunc(h.ptr, base(p), C.int(len(p))) != 0 {
+		return len(p), nil
+	}
+
+	runtime.KeepAlive(h) // Ensure the hash object is not garbage-collected
+
+	return 0, err
+}
+
+func (h *evpHash) WriteString(s string) (n int, err error) {
+	p := []byte(s)
```

Comment:
don't cast it to a byte slice (that can allocate), better use `p := unsafe.Pointer(unsafe.StringData(s))`.

Reply:
This methods are copy pasted, this is the approach commoncrypto implementations also using, lets merge it with same functionality and then in subsequent PR fix both of their implementations (I don't want to fix commoncrypto functions implementations in this PR)